File: src/Organon/Tfl/Hole.hs
module Organon.Tfl.Hole
  ( SignedTermH (..),
    StatementH (..),
    InferenceH (..),
  )
where

import Organon.Tfl.Types

-- | A signed term that may be a hole.
data SignedTermH
  = ConcreteSTH SignedTerm
  | HoleSTH
  deriving stock (Eq, Show)

-- | A statement that may contain holes, or be entirely unknown.
data StatementH
  = StmtH [SignedTermH]
  | WholeStmtH
  deriving stock (Eq, Show)

-- | An inference with possible holes in statements.
data InferenceH = InfH
  { premises :: [StatementH],
    conclusion :: StatementH
  }
  deriving stock (Eq, Show)

File: web/src/core/tfl/hole.ts
// TFL hole types — ported from Organon.Tfl.Hole

import type { SignedTerm } from "./types";

export type SignedTermH =
  | { readonly tag: "ConcreteSTH"; readonly st: SignedTerm }
  | { readonly tag: "HoleSTH" };

export type StatementH =
  | { readonly tag: "StmtH"; readonly terms: readonly SignedTermH[] }
  | { readonly tag: "WholeStmtH" };

export type InferenceH = {
  readonly premises: readonly StatementH[];
  readonly conclusion: StatementH;
};
